src/shared/Core/Trace2Exception.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.ComponentModel;
+using System.IO;
+using GitCredentialManager.Authentication.OAuth;
+using GitCredentialManager.Interop;
+
+namespace GitCredentialManager;
+
+public class Trace2Exception : Exception
+{
+    public Trace2Exception(ITrace2 trace2, string message) : base(message)
+    {
```

Comment:
Should all of these be guarded with `if (trace2 != null)` ??

Reply:
I don't think so - by virtue of having the `Trace2` instance to pass to this, we know that it's been created. It may not be initialized yet, but we guard against that in the `WriteError` method.